### PR TITLE
Removing own interfaces from the list of nodes

### PR DESCRIPTION
**Pull request by sorenmat**
This is done to avoid trying to connect to your self
when joining a cluster

### CODE REVIEW

File: src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java [hunk 1]
```diff
@@ -140,6 +158,53 @@ private Resolver buildResolver(Settings settings) {
...
+            logger.error(e1.toString());
+        }
+        while (e.hasMoreElements()) {
+            NetworkInterface n = (NetworkInterface) e.nextElement();
+            Enumeration ee = n.getInetAddresses();
+            while (ee.hasMoreElements()) {
+                InetAddress i = (InetAddress) ee.nextElement();
+                logger.info("Local interface address: " + i.getHostAddress());
+                if (i.getHostAddress().equals((ip))) {
+                    logger.info("Filtered out interface address " + ip + " because it was an local interface.");
```
Comment (tnn) [inline src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java]: .debug?

File: src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java [hunk 2]
```diff
@@ -140,6 +158,53 @@ private Resolver buildResolver(Settings settings) {
...
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    protected List<Record> filterOutOwnRecord(Record[] records) {
+        List<Record> discoNodes = Lists.newArrayList();
+        logger.info("Trying to filter out localhost records from: ", Joiner.on(", ").join(discoNodes));
```
Comment (tnn) [inline src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java]: Debug? Isn't it normal behaviour that it's filtering out itself?

File: src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java [hunk 3]
```diff
@@ -140,6 +158,53 @@ private Resolver buildResolver(Settings settings) {
...
+        for (Record record : records) {
+            SRVRecord srv = (SRVRecord) record;
+
+            String hostname = srv.getTarget().toString().replaceFirst("\\.$", "");
```
Comment (tnn) [inline src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java]: Maybe abstract this out into a static method with a name indicating why the string replace needs to happen? E.g. std RFC DNS name to xyz

File: src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java [hunk 4]
```diff
@@ -140,6 +158,53 @@ private Resolver buildResolver(Settings settings) {
...
+                hostname = hostname.replace(settings.get(DISCOVERY_SRV_CONSULPOSTFIX), "");
+            }
+            try {
+                InetAddress address = InetAddress.getByName(hostname);
+                logger.info("Resolved {} to {}", hostname, address.getHostAddress());
+                if (!isLocalIP(address.getHostAddress())) {
+                    discoNodes.add(record);
+                }
+            } catch (UnknownHostException e) {
+                logger.error(e.toString());
```
Comment (tnn) [inline src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java]: Maybe indicate that the node was not added to list

File: src/test/org/elasticsearch/discovery/srv/SrvUnicastHostsProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Test
+    public void non_local_address_returns_false() {
+        SrvUnicastHostsProvider provider = new SrvUnicastHostsProvider(ImmutableSettings.EMPTY, null, null);
+        boolean isLocal = provider.isLocalIP("255.255.255.255");
+        assertFalse(isLocal);
+    }
+
+    @Test
+    public void non_local_address_returns_fals123e() {
```
Comment (tnn) [inline src/test/org/elasticsearch/discovery/srv/SrvUnicastHostsProviderTest.java]: `fals123e`? :)

File: src/test/org/elasticsearch/discovery/srv/SrvUnicastHostsProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    public void isTCP() {
+        SrvUnicastHostsProvider provider = new SrvUnicastHostsProvider(ImmutableSettings.EMPTY, null, null);
+        boolean isTCP = provider.isTCP("tcp");
+        assertTrue(isTCP);
+
+        boolean isUDP = provider.isTCP("udp");
+        assertFalse(isUDP);
+    }
+
+}
```
Comment (tnn) [inline src/test/org/elasticsearch/discovery/srv/SrvUnicastHostsProviderTest.java]: newline

File: src/test/org/elasticsearch/discovery/srv/SrvUnicastHostsProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+            return new SRVRecord(hostname, Type.SRV, DClass.ANY, 30, 1, 80, target);
+        } catch (TextParseException e) {
+            e.printStackTrace();
+        }
+        throw new RuntimeException();
+    }
+
+    @Test
+    public void isTCP() {
```
Comment (tnn) [inline src/test/org/elasticsearch/discovery/srv/SrvUnicastHostsProviderTest.java]: descriptive name? :)

File: src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java [hunk 8]
```diff
@@ -118,6 +130,10 @@ private Resolver buildResolver(Settings settings) {
         return parent_resolver;
     }
 
+    protected boolean isTCP(String protocol) {
```
Comment (tnn) [inline src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java]: static?

File: src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java [hunk 9]
```diff
@@ -118,6 +130,10 @@ private Resolver buildResolver(Settings settings) {
...
+        return "tcp".equals(protocol);
```
Comment (tnn) [inline src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java]: equalsIgnoreCase maybe?

**Comment by sorenmat** [comment]
Replaced by #3

**Pull request closed without merging**